app/models/balance.rb

```diff
@@ -0,0 +1,87 @@
+# This describes disk quota utilization for a given user and volume
+require 'open-uri'
+
+class Balance
+  class InvalidBalanceFile < StandardError; end
+
+  attr_reader :user, :project, :value, :project_type, :unit, :updated_at
+
+  class << self
+
+    # Get balance objects only for requested user in JSON file(s)
+    #
+    # KeyError and JSON::ParserErrors shall be non-fatal errors
+    def find(balance_path, user)
+      raw = open(balance_path).read
+      raise InvalidBalanceFile.new("No content returned when attempting to read balance file") if raw.nil? || raw.empty?
+
+      # Attempt to parse raw JSON into an object
+      json = JSON.parse(raw)
+      raise InvalidBalanceFile.new("Balance file expected to be a JSON object with balances array section") unless json.is_a?(Hash) && json["balances"].respond_to?(:each)
+
+      #FIXME: any validation of the structure here? otherwise we don't need the complexity of the code below
+      # until we have more than one balance version schema, which we do not
+      # so assume version is 1
+      build_balances(json["balances"], json["timestamp"], json["config"], user)
+    rescue StandardError => e
+      Rails.logger.error("Error #{e.class} when reading and parsing balance file #{balance_path} for user #{user}: #{e.message}")
+      []
+    end
+
+    private
+
+    # Parse JSON object using version 1 formatting
+    def build_balances(balance_hashes, updated_at, config, user)
+      balances = []
+      balance_hashes.each do |balance|
+        balance = balance.to_h.compact.symbolize_keys
+        config = config.to_h.compact.symbolize_keys
+        next unless user == balance[:user]
+        balances << Balance.new(
+          user: balance.fetch(:user).to_s,
+          project: balance.fetch(:project, nil).to_s,
+          value: balance.fetch(:value).to_i,
+          project_type: config.fetch(:project_type, nil).to_s,
+          unit: config[:unit].to_s,
+          updated_at: Time.at(updated_at.to_i),
+        )
+      end
+      balances
+    end
+  end
+
+  # @param params [#to_h] list of parameters that define balance object
+  # @option params [#to_s] :user user name
+  # @option params [#to_s] :project project name
+  # @option params [#to_i] :value balance value
+  # @option params [#to_s] :project_type project type
+  # @option params [#to_s] :unit value unit
+  # @option params [#to_i] :updated_at time when balance was generated
+  def initialize(params)
+    params = params.to_h.compact.symbolize_keys
+
+    @user = params.fetch(:user).to_s
+    @project = params.fetch(:project, nil).to_s
+    @value = params.fetch(:value).to_i
+    @project_type = params.fetch(:project_type, nil).to_s
+    @unit = params.fetch(:unit).to_s
+    @updated_at = Time.at(params.fetch(:updated_at).to_i)
+  end
+
+  def balance_object
+    return @project if @project.present?
+    @user
```

Comment:
Can use the new [presence](https://guides.rubyonrails.org/active_support_core_extensions.html#presence) here:

```ruby
def balance_object
  project.presence || user
end
```

Reply:
Change made.